### PR TITLE
reduxjs/toolkit and tanstack-query (and remove custom code where possible!)

### LINKED ISSUE #6007: [Tracking] Navigation
Adding the new navigator APIs to the views.

iOS audit
- [x] SIS
  - [x] Balances
  - [x] Course search
  - [x] TES
  - [x] Open jobs
- [x] stoPrint
  - [x] Print jobs
  - [x] Release
  - [x] Printer list
  - [x] Error views
- [x] Menus
- [x] News
- [x] Building hours
  - [x] List
  - [x] Detail
  - [x] Favorite
  - [x] Report a problem
  - [x] Edit schedule
- [x] Calendar
  - [x] St Olaf
  - [x] Oleville
  - [x] Northfield
  - [x] Add to calendar
  - [x] Event Detail
- [x] Directory (link)
- [x] Directory (native)
  - [x] List
  - [x] Detail
- [x] Streaming media
  - [x] Events
  - [x] Webcams
  - [x] KSTO
  - [x] KRLX
  - [x] Show schedule    
- [x] Campus map (link)
- [x] Important contacts
  - [x] List
  - [x] Detail
- [x] Transportation
  - [x] Red
  - [x] Blue
  - [x] Olesgo
  - [x] Other modes
  - [x] Other modes detail
  - [x] Map
- [x] Campus Dictionary
  - [x] List
  - [x] Detail
  - [x] Search
  - [x] Report
- [x] Student Orgs
  - [x] List
  - [x] Detail
  - [x] Search
- [x] Moodle (link)
- [x] Oleville (link)
- [x] Report a problem
- [x] Settings
  - [x] Login
  - [x] Contact us
  - [x] FAQs
  - [x] Reset everything
  - [x] Change app icon
  - [x] Credits
  - [x] Privacy policy
  - [x] Legal
  - [x] Contributing (link)
- [x] Debug settings
  - [x] API tester
  - [x] Bon appetit picker
  - [x] Debug
  - [x] Send a sentry message
  - [x] Send a sentry exception
  - [x] Server url

### LINKED ISSUE #6739: Pause menu loading error
> A problem occured while loading: Error: `input` must not begin with a slash when using `prefixUrl`

### LINKED ISSUE #6740: API tester crashes when switching to "parsed" mode
I guess I haven't set up keyPath through the navigation correctly yet

### LINKED ISSUE #6741: Balances no longer shows N/A when logged out
It also doesn't show the LogIn button… 

Stoprint does, though. Interesting.

### LINKED ISSUE #6742: Menu filters on cors_icon (an object) allow through items with no cors_icons at all
In other words, meat dishes are showing up for `(V)`

### LINKED ISSUE #6744: All calendar views say "no events," and then something about "undefined" if you do a pull-to-refresh
I should go through and take out the "error instanceof Error" checks…

### LINKED ISSUE #6745: Student Work fails with a 504
A problem occured while loading: HTTPError: Request failed with status code 504

---

It didn't fail last night-is the endpoint working?

### LINKED ISSUE #6746: Directory fails to load
iOS 
> Something between you and the directory is having problems. Try again in a minute or two?

Android
> Error: JSON Parse error: Unexpected token: <

### LINKED ISSUE #6747: Recent Searches/Filters are not updated in course catalog view

### PR DESCRIPTION
**Pull request by hawkrives**
Closes https://github.com/StoDevX/AAO-React-Native/issues/6007
Closes https://github.com/StoDevX/AAO-React-Native/discussions/6116

Steps:
1. add dep on reduxjs/toolkit
 and start replacing our code with redux toolkit
2. add dep on redux-persist to automatically load/store data for redux
3. add dep on tanstack-query for all network requests and async state management
4. ???
5. Profit!

---

By using redux-toolkit (and redux-persist), we can automatically load/store data from AsyncStorage.

However, we probably shouldn't be using Redux to cache course data and printer data, esp. since that's only used by one view each.

I have moved all network requests to [Tanstack Query](https://tanstack.com/query/latest), which is an "async data manager", leaving only user-controlled data in Redux. (I also use [`ky`](https://github.com/sindresorhus/ky), a fetch wrapper, which has a more ergonomic API.)

I've also configured Tanstack Query to cache all network requests for ~1 day, which gets us actual offline support for the first time!

Finally, I've been porting screens to hooks as I've encountered them, so that they can `useQuery`. I've removed several vestigial "connected" components. 

Finally, I fixed a bunch of type errors with the Filters stuff, after I remembered how it worked. 

Finally, I reworked networking APIs where it made sense, so I think the olecard and login stuff is a lot nicer now. 

---

To Do:

- [x] ccc apis
- [x] course search
- [x] balances
- [x] login/auth
- [x] stoprint

Known Issues:

- [ ] API debugger fails in Parsed mode
- [ ] Menu filters on cors_icon (an object) allow through items with no cors_icons at all (meat dishes are showing up for (V))
- [ ] Balances no longer shows N/A when not logged in

### CODE REVIEW

File: source/views/directory/query.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+let directory = ky.create({prefixUrl: 'https://www.stolaf.edu/directory'})
+
+const getDirectoryParams = (query: string, type: DirectorySearchTypeEnum) => {
+	let p = new URLSearchParams()
+	p.set('format', 'json')
+
+	switch (type) {
+		case 'department':
+			p.set('department', query.trim())
```
Comment (drewvolz) [inline source/views/directory/query.ts]: What are your thoughts on calling query.trim() once before the switch/case?

File: source/views/building-hours/query.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import {useSelector} from 'react-redux'
+import {queryClient} from '../../init/tanstack-query'
+import {selectFavoriteBuildings} from '../../redux'
+import {BuildingType} from './types'
+
+export const keys = {
+	all: ['buildings'] as const,
+}
+
+queryClient.setQueryData(keys.all, require('../../docs/building-hours.json'))
```
Comment (drewvolz) [inline source/views/building-hours/query.ts]: Eslint

> Require statement not part of import statement

File: modules/ccc-calendar/query.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
+import {client} from '@frogpond/api'
+import {EventType} from '@frogpond/event-type'
+import {useQuery} from '@tanstack/react-query'
+import moment from 'moment'
+import {GoogleCalendar, IcsCalendar, NamedCalendar, ReasonCalendar, type Calendar} from './types'
```
Comment (drewvolz) [inline modules/ccc-calendar/query.ts]: Do we want to remove `type Calendar`?

File: modules/ccc-calendar/index.tsx [hunk 4]
```diff
@@ -1,113 +1,45 @@
...
-		this.setState(() => ({refreshing: true}))
-		await this.getEvents(true)
-		this.setState(() => ({refreshing: false}))
-	}
-
-	render(): React.ReactNode {
-		if (!this.state.initialLoadComplete) {
-			return <LoadingView />
-		}
+export function CccCalendarView(props: Props) {
```
Comment (drewvolz) [inline modules/ccc-calendar/index.tsx]: I suppose we can add return types for the five functions eslint has found

**Review by drewvolz — APPROVED** [review]
Wow, what an improvement. Great job!

File: source/views/directory/query.ts (hunk 1 above)
Comment (hawkrives) [reply source/views/directory/query.ts]: I'd be down for that! They do all share "query", after all

**Review by drewvolz — APPROVED** [review]
This is stunning. Very nicely done, Hawken!

**Pull request merged**